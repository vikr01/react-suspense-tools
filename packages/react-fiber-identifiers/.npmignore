src/
tsup.config.js
typescript.config.json
.gitignore
.npmignore
__tests__
node_modules/
