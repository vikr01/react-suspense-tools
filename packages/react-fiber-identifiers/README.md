# react-fiber-identifiers

Utilities for identifying React fibers.

## react-fiber-identifiers/is-error-boundary

Returns `true` if fiber is for an error boundary, `false` if not.

```javascript
import isErrorBoundary from 'react-fiber-idenfitifers/is-error-boundary';

isErrorBoundary(fiberFrom_ErrorBoundary); // true
isErrorBoundary(fiberFrom_Non_ErrorBoundary); // false
```

## react-fiber-identifiers/is-suspense

Returns `true` if fiber is for a `React.Suspense`, `false` if not.

```javascript
import isSuspense from 'react-fiber-idenfitifers/is-error-boundary';

isSuspense(fiberFrom_ReactSuspense); // true
isSuspense(fiberFrom_AnythingElse); // false
```

## react-fiber-identifiers/get-unique-identifier

Returns a `string` for a fiber's "mount" state (`key`, or index within parent fiber node of `key` isn't specified in props).

```javascript
import getUniqueIdentifier from 'react-fiber-identifiers/get-unique-identifier';
```

## react-fiber-identifiers/testing/create-fiber

For jest environments, to create a React fiber.

```jsx
import createFiber from 'react-fiber-identifiers/testing/create-fiber';

createFiber(<Foo />); // creates a fiber from `Foo`
```