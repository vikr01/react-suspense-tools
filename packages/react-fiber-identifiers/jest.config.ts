import type { Config } from "@jest/types";
import path from "path";
import baseConfig from "../../jest.config.base";

const config: Config.InitialOptions = {
  ...baseConfig,

  moduleNameMapper: {
    ...baseConfig.moduleNameMapper,
    // "^shared$": "@gitpkg/react-shared",
  },

  setupFiles: [
    ...(baseConfig.setupFiles ?? []),
    require.resolve("./testing/react-github-setup.jest.js"),
  ],

  transform: {
    ...baseConfig.transform,

    // strip flow types out of react source code pulled directly from github
    "@gitpkg/react-.*": require.resolve("../../testing/deleteFlowTypes.babel"),
    [path.dirname(require.resolve("shared/package.json"))]: require.resolve(
      "../../testing/deleteFlowTypes.babel",
    ),
  },

  transformIgnorePatterns: [
    ...(baseConfig.transformIgnorePatterns ?? []),
    "/node_modules/(?!@gitpkg/react-.*)/",
    "/node_modules/(?!shared)/",
  ],
};

export default config;
