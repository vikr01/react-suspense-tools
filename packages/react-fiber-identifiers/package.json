{
  "name": "react-fiber-identifiers",
  "version": "0.0.0",
  "exports": {
    "./get-unique-identifier": "./build/getUniqueIdentifier.js",
    "./is-error-boundary": "./build/isErrorBoundary.js",
    "./is-suspense": "./build/isSuspense.js"
  },
  "scripts": {
    "build": "tsup",
    "test": "jest"
  },
  "author": {
    "name": "Vik R",
    "email": "vik120@icloud.com",
    "url": "http://github.com/vikr01/"
  },
  "license": "ISC",
  "peerDependencies": {
    "react": "^19.0.0"
  },
  "dependencies": {
    "nanoid": "^5.1.5"
  },
  "devDependencies": {
    "@gitpkg/react-reconciler": "https://gitpkg.vercel.app/facebook/react/packages/react-reconciler?v19.0.0",
    "@gitpkg/react-shared": "https://gitpkg.vercel.app/facebook/react/packages/shared?v19.0.0"
  }
}
