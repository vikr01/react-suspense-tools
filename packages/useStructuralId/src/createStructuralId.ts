import { traverseFiber, type Fiber as FineFiber } from "its-fine";
import type { Fiber } from "react-reconciler";
import getUniqueIdentifier from "react-fiber-identifiers/get-unique-identifier";

export type Selector = null | Parameters<typeof traverseFiber>[2];
export type StructuralId = string;
export type StopNode = Fiber | null;
export type { Fiber };

function createStructuralId(
  fiber: Fiber,
  selector: Selector,
): [StructuralId, StopNode] {
  const uniqueKeys: Array<ReturnType<typeof getUniqueIdentifier>> = [];

  const stopNode =
    traverseFiber(fiber, true, function (node: FineFiber, ...args) {
      uniqueKeys.push(getUniqueIdentifier(node));
      const res = selector?.(node, ...args);
      return res;
    }) ?? null;

  return [uniqueKeys.join(","), stopNode];
}

export default function creatStructuralIdWithHookCallIndex(
  num: number,
  ...rest: Parameters<typeof createStructuralId>
): [StructuralId, StopNode] {
  const [structuralidWithoutHookCall, stopNode] = createStructuralId(...rest);
  return [`${num}\\${structuralidWithoutHookCall}`, stopNode];
}
