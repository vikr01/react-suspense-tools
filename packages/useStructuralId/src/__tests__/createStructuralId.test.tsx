import { cleanup, render } from "@testing-library/react";
import * as React from "react";
import createStructuralId from "../createStructuralId";
import { clear } from "react-fiber-identifiers/get-unique-identifier";
import type { Fiber } from "../createStructuralId";

Object.defineProperty(global, "__DEV_STRUCTURAL_ID_DEBUG__", {
  value: false,
  writable: true,
  configurable: true,
});

type ChildrenOnlyProps = {
  children: React.ReactNode;
};

describe("createStructuralId", () => {
  let fiber: null | Fiber = null;
  let fiber3: null | Fiber = null;
  afterEach(() => {
    clear();
    cleanup();
    fiber = null;
  });

  function Component0({ children }: ChildrenOnlyProps) {
    fiber =
      // eslint-disable-next-line @typescript-eslint/no-explicit-any, prettier/prettier
        (React as any)?.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE?.A?.getOwner?.();
    return children;
  }
  const Component1 = ({ children }: ChildrenOnlyProps) => children;
  class Component2 extends React.Component<ChildrenOnlyProps> {
    render() {
      return this.props.children;
    }
  }
  const Component3 = React.forwardRef(function Component3(
    { children }: ChildrenOnlyProps,
    // silence react complaining there was no second parameter

    _ref,
  ) {
    fiber3 =
      // eslint-disable-next-line @typescript-eslint/no-explicit-any, prettier/prettier
      (React as any)?.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE?.A?.getOwner?.();

    return children;
  });
  const Component4 = React.Fragment;
  const Component5 = React.memo(({ children }: ChildrenOnlyProps) => children);
  const Component6 = ({ children }: ChildrenOnlyProps) => <a>{children}</a>;
  const Component7 = React.Fragment;

  it("genereates a string from elements with various element types", async () => {
    const components = [
      Component0,
      Component1,
      Component2,
      Component3,
      Component4,
      Component5,
      Component6,
      Component7,
    ] as const;

    const renderNode = components.reduce(
      (node: React.ReactNode, Component) => <Component>{node}</Component>,
      <div data-testid="leaf" />,
    );

    render(renderNode);

    expect(fiber).not.toBeNull();
    const [id] = createStructuralId(151, fiber as Fiber, null);

    expect(id).toMatchSnapshot();
  });

  it("properly reuses ids of component pointers", () => {
    const components = [
      [Component0, null],
      [Component2, "foobar"],
      [Component3, "bazfoo"],
      [Component4, 10],
      [Component7, 15],
    ] as const;

    const renderNode = components.reduce(
      (node: React.ReactNode, [Component, key]) => (
        <Component key={key}>{node}</Component>
      ),
      <div data-testid="leaf2" />,
    );

    render(renderNode);

    expect(fiber).not.toBeNull();

    const structuralNodes = createStructuralId(
      505,
      fiber as Fiber,
      (node: Fiber) => node.elementType === Component3,
    )[0]
      .split("\\")[1]
      .split(",");

    const component3UniqueId =
      structuralNodes[structuralNodes.length - 1].split(":")[0];

    const key = "xyz";
    render(<Component3 children={null} key={key} />);

    const component3AsElementUniqueId = createStructuralId(
      505,
      fiber3 as Fiber,
      null,
    )[0]
      .split("\\")[1]
      .split(",")[0];

    expect(component3AsElementUniqueId).toBe(`${component3UniqueId}:s[${key}]`);
  });

  it("makes readable ids when the flag is on", () => {
    Object.defineProperty(global, "__DEV_STRUCTURAL_ID_DEBUG__", {
      value: true,
      writable: true,
      configurable: true,
    });

    const components = [
      [Component0, null],
      [Component1, 39],
      [Component2, null],
      [Component3, null],
      [Component4, "foobar"],
      [Component5, null],
      [Component6, null],
      [Component7, null],
    ] as const;

    const renderNode = components.reduce(
      (node: React.ReactNode, [Component, key]) => (
        <Component key={key}>{node}</Component>
      ),
      <div data-testid="leaf3" />,
    );

    render(renderNode);

    expect(fiber).not.toBeNull();

    const [id] = createStructuralId(22, fiber as Fiber, null);

    expect(id).toMatchSnapshot();
  });

  it("gets the appropriate stop node", () => {
    const comp5Key = 10;
    const components = [
      [Component0, null],
      [Component7, 15],
      [Component3, "bazfoo"],
      [Component5, comp5Key],
      [Component2, "foobar"],
    ] as const;

    const renderNode = components.reduce(
      (node: React.ReactNode, [Component, key]) => (
        <Component key={key}>{node}</Component>
      ),
      <div data-testid="leaf2" />,
    );

    render(renderNode);

    expect(fiber).not.toBeNull();
    const res = createStructuralId(
      10,
      fiber as Fiber,
      (node: Fiber) => node.elementType === Component5,
    );

    const [, stopNode] = res;

    expect(stopNode?.elementType).toBe(Component5);

    expect(Number(stopNode?.key)).toBe(comp5Key);
  });
});
