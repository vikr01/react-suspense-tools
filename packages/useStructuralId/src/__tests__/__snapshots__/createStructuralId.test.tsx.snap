// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStructuralId genereates a string from elements with various element types 1`] = `"151\\000000:n[0],000001:n[0],000002:n[0],000003:n[0],000004:n[0],a:n[0],000005:n[0],Unknown:n[0]"`;

exports[`createStructuralId makes readable ids when the flag is on 1`] = `"22\\Component0:s[null],Component1:s[39],Component2:s[null],Fallback:s[null],Unknown:s[foobar],Fallback:s[null],a:n[0],Component6:s[null],Unknown:s[null],Unknown:n[0]"`;
